### PR TITLE
decompose test into two parts

### PR DESCRIPTION
**Pull request by tballmsft**

### CODE REVIEW

File: docs/src/components/ServiceUnitTest.tsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import { JDService } from '../../../src/jdom/service';
+
+enum TestStatus {
+    Inactive,
+    Active,
+    Passed,
+    Failed,
+}
+
+export function ServiceUnitTest(props: {serviceInstance: JDService, test: jdtest.UnitTest}) {
```
Comment (pelikhan) [inline docs/src/components/ServiceUnitTest.tsx:22]: make sure to format your code using ``Shift+Alt+F``

Comment (pelikhan) [inline docs/src/components/ServiceUnitTest.tsx:22]: add default -> ``export default function ....``

File: docs/src/components/ServiceUnitTest.tsx [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    const { serviceInstance, test } = props
+    const [ testStatus, setTestStatus ] = useState<TestStatus>(TestStatus.Inactive)
+    return (
+        <Step key={test.description}>
```
Comment (pelikhan) [inline docs/src/components/ServiceUnitTest.tsx:26]: you only need key when creating a list of React nodes

File: docs/src/components/ServiceUnitTest.tsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        <StepLabel>{test.description}</StepLabel>
```
Comment (pelikhan) [inline docs/src/components/ServiceUnitTest.tsx:27]: do you expect ``test.description`` to be markdown? you can use ``<Markdown ... />``

**Comment by pelikhan** [comment]
Use a grid to control the size of elements ( https://material-ui.com/components/grid/ ) so that you can control the size for multiple screen size.

File: docs/src/components/ServiceUnitTest.tsx (hunk 2 above)
Comment (tballmsft) [reply docs/src/components/ServiceUnitTest.tsx]: yeah, this was in a loop. has been refactored

File: docs/src/components/ServiceUnitTest.tsx (hunk 1 above)
Comment (tballmsft) [reply docs/src/components/ServiceUnitTest.tsx]: done.

File: docs/src/components/ServiceTest.tsx [hunk 4]
```diff
@@ -1,88 +1,46 @@
...
+// - allow to select a test to run it (only one active test at a time)
+// - an active test has an active command
 
 export default function ServiceTest(props: { serviceSpec: jdspec.ServiceSpec }) {
     const { serviceSpec } = props
     const { classIdentifier: serviceClass } = serviceSpec
     const { bus } = useContext<JacdacContextProps>(JacdacContext)
-    const classes = useStyles();
-    const [selectedService, setSelectedService] = useState<JDService>(undefined);
+    const [selectedServiceInstance, setSelectedService] = useState<JDService>(undefined);
```
Comment (pelikhan) [inline docs/src/components/ServiceTest.tsx:26]: i typically call those "service", not "serviceInstance", why the rename?

File: docs/src/components/ServiceTest.tsx [hunk 5]
```diff
@@ -1,88 +1,46 @@
...
     const handleSelect = (service: JDService) => () => {
         setSelectedService(service);
-        setActiveStep(1);
+        handleNext();
     }
+    const handleStartTest = (t: number) => () => {
+        setActiveTest(() => t);
+    };
     const handleNext = () => {
         setActiveStep((prevActiveStep) => prevActiveStep + 1);
```
Comment (pelikhan) [inline docs/src/components/ServiceTest.tsx:45]: cool!

File: docs/src/components/ServiceTest.tsx [hunk 6]
```diff
@@ -92,65 +50,69 @@ export default function ServiceTest(props: { serviceSpec: jdspec.ServiceSpec })
     const handleReset = () => {
         setActiveStep(0);
     };
+    const testFinished = (status: TestStatus) => {
+        setActiveStep((prevActiveStep) => prevActiveStep + 1);
```
Comment (pelikhan) [inline docs/src/components/ServiceTest.tsx:54]: handleNext?

File: docs/src/components/ServiceTest.tsx [hunk 7]
```diff
@@ -92,65 +50,69 @@ export default function ServiceTest(props: { serviceSpec: jdspec.ServiceSpec })
...
-                                <li>Use the <strong>packet console</strong> to monitor packets on the bus</li>
-                                <li>Check the class identifier in your annoucement packets</li>
-                            </ul></Alert>}
-                    </StepContent>
-                </Step>
-                {tests.map(({ id, label, description }) => (
-                    <Step key={id}>
-                        <StepLabel>Test {id}: {label}</StepLabel>
+        <Grid container spacing={2}>
+            <Grid>
```
Comment (pelikhan) [inline docs/src/components/ServiceTest.tsx:59]: <Grid item>

File: docs/src/components/ServiceTest.tsx [hunk 8]
```diff
@@ -92,65 +50,69 @@ export default function ServiceTest(props: { serviceSpec: jdspec.ServiceSpec })
...
-                                        color="primary"
-                                        onClick={handleNext}
-                                        className={classes.button}
-                                    >{activeStep === stepLength - 1 ? 'Finish' : 'Next'}
-                                    </Button>
-                                </div>
-                            </div>
+                            {!!serviceInstances.length && <Grid container spacing={2}>
+                                {serviceInstances.map(service => <Grid item {...gridBreakpoints}>
+                                    <Card key={service.id}>
```
Comment (pelikhan) [inline docs/src/components/ServiceTest.tsx:67]: ``key`` needs to be on the <Grid element

File: docs/src/components/ServiceTest.tsx [hunk 9]
```diff
@@ -92,65 +50,69 @@ export default function ServiceTest(props: { serviceSpec: jdspec.ServiceSpec })
...
-                                    <Button
-                                        variant="contained"
-                                        color="primary"
-                                        onClick={handleNext}
-                                        className={classes.button}
-                                    >{activeStep === stepLength - 1 ? 'Finish' : 'Next'}
-                                    </Button>
-                                </div>
-                            </div>
+                            {!!serviceInstances.length && <Grid container spacing={2}>
```
Comment (pelikhan) [inline docs/src/components/ServiceTest.tsx:65]: ``?.length`` just for safety

File: docs/src/components/ServiceTest.tsx [hunk 10]
```diff
@@ -92,65 +50,69 @@ export default function ServiceTest(props: { serviceSpec: jdspec.ServiceSpec })
...
+                                        <CardActions>
+                                            <Button variant="contained" color="primary" onClick={handleSelect(service)}>Select</Button>
+                                        </CardActions>
+                                    </Card>
+                                </Grid>)}
+                            </Grid>}
+                            {!serviceInstances.length && <Alert severity="info">Not seeing your device? Try some of the following.
+                        <ul>
+                                    <li>Check that your device is connected</li>
+                                    <li>Use the <strong>packet console</strong> to monitor packets on the bus</li>
```
Comment (pelikhan) [inline docs/src/components/ServiceTest.tsx:78]: i mean to show the icon here one day

File: docs/src/components/ServiceTest.tsx [hunk 11]
```diff
@@ -92,65 +50,69 @@ export default function ServiceTest(props: { serviceSpec: jdspec.ServiceSpec })
...
+                    <Paper square elevation={0} >
+                        <Typography>All steps completed - you're finished</Typography>
+                        <Button onClick={handleReset} >
+                            Reset
+                </Button>
+                    </Paper>
+                )}
+            </Grid>
+            {selectedServiceInstance && <DashbardDeviceItem
+                key={selectedServiceInstance.device.id}
```
Comment (pelikhan) [inline docs/src/components/ServiceTest.tsx:110]: not needed, only for collections

File: docs/src/components/ServiceUnitTest.tsx [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    })
+    function closeTestStep(index: number) {
+        currTestStep.length = currLength;
+        testSteps.push(currTestStep);
+        currLength = 0;
+        currTestStep = { start: index, length: undefined };
+    }
+}
+
+export default function ServiceUnitTest(props: { serviceInstance: JDService, test: jdtest.UnitTest, finished: (status: TestStatus) => void }) {
```
Comment (pelikhan) [inline docs/src/components/ServiceUnitTest.tsx:43]: finished -> onFinished

File: docs/src/components/ServiceUnitTest.tsx [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+    const { serviceInstance, test, finished } = props
+    const [activeCommand, setActiveCommand] = useState(0);
+    const handleNext = () => {
+        setActiveCommand((prevActiveStep) => prevActiveStep + 1);
```
Comment (pelikhan) [inline docs/src/components/ServiceUnitTest.tsx:47]: preActiveStep -> prevActiveCommand or just prev

File: docs/src/components/ServiceUnitTest.tsx [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+    };
+    const handleClose = (status: TestStatus) => () => {
+        finished(status);
+    };
+    return (<Grid>
```
Comment (pelikhan) [inline docs/src/components/ServiceUnitTest.tsx:52]: <Grid item>

File: docs/src/components/ServiceUnitTest.tsx [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+        <Stepper activeStep={activeCommand} orientation="vertical">
+            {test.commands.map((cmd, index) => (commandToUI(index, cmd)))}
+        </Stepper>
+    </Grid>);
+
+    function exprToString(token: jdtest.ServiceTestToken) {
+        return (token.js ? token.js : "") + (token?.const ? token.const.toString() : "") + (token.id ? token.id : "");
+    }
+    function commandToUI(index: number, cmd: jdtest.ServiceTestCommand) {
```
Comment (pelikhan) [inline docs/src/components/ServiceUnitTest.tsx:61]: move up and convert to 
```
const Command = (props: { .... }) {
}
```
then do 
```
<Command index={index} ... />
```

Comment (pelikhan) [reply docs/src/components/ServiceUnitTest.tsx:61]: I'd recommend moving it out in its own component; it'll make things cleaner.

**Review by pelikhan — CHANGES_REQUESTED** [review]
This is shaping up!

File: docs/src/components/ServiceTest.tsx (hunk 11 above)
Comment (tballmsft) [reply docs/src/components/ServiceTest.tsx:110]: We need to show the device twin so when the user steps through the test, they see changes.

**Comment by pelikhan** [comment]
You don’t need the key attribute

________________________________

File: docs/src/components/ServiceUnitTest.tsx (hunk 15 above)
Comment (tballmsft) [reply docs/src/components/ServiceUnitTest.tsx:61]: inlined for now until it gets big enough to move out.

File: docs/src/components/ServiceTest.tsx (hunk 7 above)
Comment (tballmsft) [reply docs/src/components/ServiceTest.tsx:59]: what's the comment here?

File: docs/src/components/ServiceTest.tsx (hunk 11 above)
Comment (tballmsft) [reply docs/src/components/ServiceTest.tsx:110]: oh

**Comment by pelikhan** [comment]
<Grid item>

________________________________

**Pull request merged**